Library/Homebrew/os/mac/cache_store.rb

```diff
@@ -0,0 +1,137 @@
+require "dbm"
+require "json"
+
+#
+# `DatabaseCache` acts as an interface to a persistent storage mechanism
+# residing in the `HOMEBREW_CACHE`
+#
+class DatabaseCache
+  # Users have read and write, but not execute permissions
+  DATABASE_MODE = 0666
+
+  # Opens and yields a database in read/write mode
+  #
+  # DBM::WRCREAT: Creates the database if it does not already exist
+  def initialize(name)
+    @db = DBM.open("#{HOMEBREW_CACHE}/#{name}.db", DATABASE_MODE, DBM::WRCREAT)
+    yield(@db)
+    @db.close
+  end
+end
+
+#
+# `CacheStore` provides methods to mutate and fetch data from a persistent
+# storage mechanism
+#
+class CacheStore
+  def initialize(database_cache)
+    @database_cache = database_cache
+  end
+
+  # Inserts new values or updates existing cached values to persistent storage
+  # mechanism
+  def update!(*)
+    raise NotImplementedError
+  end
+
+  # Fetches cached values in persistent storage according to the type of data
+  # stored
+  def fetch_type(*)
+    raise NotImplementedError
+  end
+
+  # Deletes data from the cache based on a condition defined in a concrete class
+  def flush_cache!
+    raise NotImplementedError
+  end
+
+  protected
+
+  attr_reader :database_cache
+
+  # Parses `DBM` stored `String` into ruby `Hash`
+  #
+  # DBM stores ruby objects as a ruby `String`. Hence, when fetching the data,
+  # to convert the ruby string back into a ruby `Hash`, the string is converted
+  # into a JSON compatible string, where it may be parsed by the JSON.parse
+  # function
+  def string_to_hash(string)
+    JSON.parse(string.gsub("=>", ":"))
+  end
+end
+
+#
+# `LinkageStore` provides methods to fetch and mutate linkage-specific data used
+# by the `brew linkage` command
+#
+class LinkageStore < CacheStore
+  HASH_LINKAGE_TYPES = [:brewed_dylibs, :reverse_links].freeze
+
+  def initialize(keg_name, database_cache)
+    @keg_name = keg_name
+    super(database_cache)
+  end
+
+  def update!(
+    path_values: {
```

Comment:
Do `path_values` and `hash_values` need to be separate named arguments here? Could instead you just use the inner key names as named parameters?